go/vt/wrangler/reparent.go

```diff
@@ -226,3 +179,117 @@ func (wr *Wrangler) ReparentTablet(ctx context.Context, tabletAlias topo.TabletA
 	rsd.Force = ti.Type == topo.TYPE_LAG_ORPHAN
 	return wr.tmc.RestartSlave(ctx, ti, rsd)
 }
+
+// InitShardMaster will make the provided tablet the master for the shard.
+func (wr *Wrangler) InitShardMaster(ctx context.Context, keyspace, shard string, masterElectTabletAlias topo.TabletAlias, force bool, waitSlaveTimeout time.Duration) error {
+	// lock the shard
+	actionNode := actionnode.ReparentShard(initShardMasterOperation, masterElectTabletAlias)
+	lockPath, err := wr.lockShard(ctx, keyspace, shard, actionNode)
+	if err != nil {
+		return err
+	}
+
+	// do the work
+	err = wr.initShardMasterLocked(ctx, keyspace, shard, masterElectTabletAlias, force, waitSlaveTimeout)
+
+	// and unlock
+	return wr.unlockShard(ctx, keyspace, shard, actionNode, lockPath, err)
+}
+
+func (wr *Wrangler) initShardMasterLocked(ctx context.Context, keyspace, shard string, masterElectTabletAlias topo.TabletAlias, force bool, waitSlaveTimeout time.Duration) error {
+	shardInfo, err := wr.ts.GetShard(keyspace, shard)
+	if err != nil {
+		return err
+	}
+
+	tabletMap, err := topo.GetTabletMapForShard(ctx, wr.ts, keyspace, shard)
+	if err != nil {
+		return err
+	}
+
+	// Check the master elect is in tabletMap
+	masterElectTabletInfo, ok := tabletMap[masterElectTabletAlias]
+	if !ok {
+		return fmt.Errorf("master-elect tablet %v is not in the shard", masterElectTabletAlias)
+	}
+
+	// Check the master is the only master is the shard, or -force was used.
+	_, masterTabletMap := topotools.SortedTabletMap(tabletMap)
+	if shardInfo.MasterAlias != masterElectTabletAlias {
+		if !force {
+			return fmt.Errorf("master-elect tablet %v is not the shard master, use -force to proceed anyway", masterElectTabletAlias)
+		}
+		wr.logger.Warningf("master-elect tablet %v is not the shard master, proceeding anyway as -force was used", masterElectTabletAlias)
+	}
+	if _, ok := masterTabletMap[masterElectTabletAlias]; !ok || len(masterTabletMap) != 1 {
+		if !force {
+			return fmt.Errorf("master-elect tablet %v is not the only master in the shard, use -force to proceed anyway", masterElectTabletAlias)
```

Comment:
SortedTabletMap will also put SCRAP tablets in the master map. If there are scrapped tablets, the user will get an error here saying "there's more than one master", which will be confusing if there actually is only one master. This error should mention that the problem could also be the presence of scrapped tablets.